### PR TITLE
[Feature] Show partitions statement support of list partition

### PR DESCRIPTION
**Pull request by hiliuxg**
issue #4035 

## What type of PR is this：
- [ ] bug
- [X] feature
- [ ] enhancement
- [ ] others

#### Create Table for single column partition

```sql
create database example_db  ;
use example_db ;

CREATE TABLE t_recharge_detail_single(
    id bigint  ,
    user_id  bigint  ,
    recharge_money decimal(32,2) , 
    province varchar ,
    dt varchar
) ENGINE=OLAP
DUPLICATE KEY(id)
PARTITION BY LIST (province) (
   PARTITION p1 VALUES IN ("beijing","chongqing"),
   PARTITION p2 VALUES IN ("shanghai","tianjing")
)
DISTRIBUTED BY HASH(id) BUCKETS 1
PROPERTIES(
    "replication_num" = "1"
)
```

#### Create Table for multi column partition  

```sql
CREATE TABLE t_recharge_detail_multi(
    id bigint  ,
    user_id  bigint  ,
    recharge_money decimal(32,2) , 
    province varchar ,
    dt varchar
) ENGINE=OLAP
DUPLICATE KEY(id)
PARTITION BY LIST (dt,province) (
   PARTITION p1 VALUES IN (("2022-04-01", "beijing"),("2022-04-01", "chongqing")),
   PARTITION p2 VALUES IN (("2022-04-01", "shanghai"))
)
DISTRIBUTED BY HASH(id) BUCKETS 1
PROPERTIES(
    "replication_num" = "1"
);
```

#### Show partitions from tableName
```sql
show partitions  from  t_recharge_detail_single  \G;
show partitions  from  t_recharge_detail_multi  \G;
``` 

#### Show result
image
image
image

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/common/proc/PartitionsProcDir.java [hunk 1]
```diff
@@ -66,30 +67,42 @@
...
-    static {
+    private void createTitleNames() {
         ImmutableList.Builder<String> builder = new ImmutableList.Builder<String>()
                 .add("PartitionId").add("PartitionName")
                 .add("VisibleVersion").add("VisibleVersionTime").add("VisibleVersionHash")
-                .add("State").add("PartitionKey").add("Range").add("DistributionKey")
+                .add("State").add("PartitionKey");
+
+        if (this.partitionType == PartitionType.LIST) {
+            builder.add("Values");
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/PartitionsProcDir.java:91]:
```suggestion
            builder.add("List");
```

Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/common/proc/PartitionsProcDir.java:91]: Better to keep it consistent with range partition.

File: fe/fe-core/src/main/java/com/starrocks/catalog/ListPartitionInfo.java [hunk 2]
```diff
@@ -152,17 +150,19 @@ private String singleListPartitionSql(OlapTable table, short tableReplicationNum
         return StringUtils.removeEnd(sb.toString(), ",\n");
     }
 
+    private String valuesToString(List<String> values) {
+        return "(" + values.stream().map(value -> "\'" + value + "\'")
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/ListPartitionInfo.java:154]: Use `LiteralExpr` values.
String values are only used for persistence.

File: fe/fe-core/src/main/java/com/starrocks/catalog/ListPartitionInfo.java [hunk 3]
```diff
@@ -177,8 +177,25 @@ private String multiListPartitionSql(OlapTable table, short tableReplicationNum)
         return StringUtils.removeEnd(sb.toString(), ",\n");
     }
 
+    private String multiValuesToString(List<List<String>> multiValues) {
+        return multiValues.stream()
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/ListPartitionInfo.java:181]: Use `LiteralExpr` values.
String values are only used for persistence.

File: fe/fe-core/src/main/java/com/starrocks/common/proc/PartitionsProcDir.java [hunk 4]
```diff
@@ -66,30 +67,42 @@
  * show [temp] partitions' detail info within a table
  */
 public class PartitionsProcDir implements ProcDirInterface {
-    public static final ImmutableList<String> TITLE_NAMES;
+    private final PartitionType partitionType;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/PartitionsProcDir.java:70]: Please add `PartitionType` in `TablesProcDir`.

**Comment by gengjun-git** [comment]
Can you give a show result example?

**Comment by hiliuxg** [comment]
> Can you give a show result example?

Thanks for review , I had paste the result

File: fe/fe-core/src/main/java/com/starrocks/catalog/ListPartitionInfo.java [hunk 5]
```diff
@@ -102,6 +137,19 @@ public static PartitionInfo read(DataInput in) throws IOException {
...
+            Map<Long, List<String>> idToValuesMap = partitionInfo.getIdToValues();
+            for (Map.Entry<Long, List<String>> entry : idToValuesMap.entrySet()) {
+                partitionInfo.setLiteralExprValues(entry.getKey(), entry.getValue());
+            }
+            Map<Long, List<List<String>>> idToMultiValuesMap = partitionInfo.getIdToMultiValues();
+            for (Map.Entry<Long, List<List<String>>> entry : idToMultiValuesMap.entrySet()) {
+                partitionInfo.setMultiLiteralExprValues(entry.getKey(), entry.getValue());
+            }
+        } catch (AnalysisException e) {
+            throw new IOException(e);
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/catalog/ListPartitionInfo.java:151]: Do not throw exception here. Just print error log. Exception in replay log will cause process exit.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 108 / 171 (63.16%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/proc/TablesProcDir.java|1|50|02.00%|
|:large_blue_circle:|com/starrocks/common/proc/PartitionsProcDir.java|29|34|85.29%|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|46|52|88.46%|
|:large_blue_circle:|com/starrocks/analysis/ListPartitionDesc.java|21|23|91.30%|
|:large_blue_circle:|com/starrocks/analysis/ShowPartitionsStmt.java|11|12|91.67%|

File: fe/fe-core/src/main/java/com/starrocks/analysis/ListPartitionDesc.java [hunk 6]
```diff
@@ -171,26 +171,33 @@ private void analyzeDuplicateValues(List<LiteralExpr> allLiteralExprValues) thro
...
-        }
-        for (MultiItemListPartitionDesc desc : this.multiListPartitionDescs){
-            long partitionId = partitionNameToId.get(desc.getPartitionName());
-            listPartitionInfo.setDataProperty(partitionId, desc.getPartitionDataProperty());
-            listPartitionInfo.setIsInMemory(partitionId, desc.isInMemory());
-            listPartitionInfo.setTabletType(partitionId, desc.getTabletType());
-            listPartitionInfo.setReplicationNum(partitionId, desc.getReplicationNum());
-            listPartitionInfo.setMultiValues(partitionId, desc.getMultiValues());
+    public PartitionInfo toPartitionInfo(List<Column> columns, Map<String, Long> partitionNameToId, boolean isTemp) throws DdlException {
+        try{
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ListPartitionDesc.java:175]: Please format the code.

File: fe/fe-core/src/main/java/com/starrocks/catalog/ListPartitionInfo.java [hunk 7]
```diff
@@ -152,17 +202,19 @@ private String singleListPartitionSql(OlapTable table, short tableReplicationNum
...
             Optional.ofNullable(table.getPartition(partitionId)).ifPresent(partition -> {
                 String partitionName = partition.getName();
                 sb.append("  PARTITION ").append(partitionName).append(" VALUES IN (");
-                String items = multiValues.stream()
-                        .map(values -> "(" + values.stream().map(value -> "\'" + value + "\'")
-                                .collect(Collectors.joining(",")) + ")")
-                        .collect(Collectors.joining(","));
+                String items = this.multiValuesToString(multiValues);
                 sb.append(items);
                 sb.append(")");
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/ListPartitionInfo.java:219]: Remove `()` here and put it into `multiValuesToString`.

Consistent with `valuesToString`.

File: fe/fe-core/src/main/java/com/starrocks/common/proc/PartitionsProcDir.java [hunk 8]
```diff
@@ -248,19 +261,9 @@ private List<List<Comparable>> getPartitionInfos() {
...
-                    for (Column column : partitionColumns) {
-                        colNames.add(column.getName());
-                    }
-                    partitionInfo.add(joiner.join(colNames));
-                    partitionInfo.add(((RangePartitionInfo) tblPartitionInfo).getRange(partitionId).toString());
-                } else {
-                    partitionInfo.add("");
-                    partitionInfo.add("");
-                }
+                // partition key , range or value
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/PartitionsProcDir.java:264]:
```suggestion
                // partition key , range or list value
```

File: fe/fe-core/src/main/java/com/starrocks/common/proc/TablesProcDir.java [hunk 9]
```diff
@@ -39,14 +41,15 @@
...
+import java.util.stream.Collectors;
 
 /*
  * SHOW PROC /dbs/dbId/
  * show table family groups' info within a db
  */
 public class TablesProcDir implements ProcDirInterface {
     public static final ImmutableList<String> TITLE_NAMES = new ImmutableList.Builder<String>()
-            .add("TableId").add("TableName").add("IndexNum").add("PartitionColumnName")
+            .add("TableId").add("TableName").add("IndexNum").add("PartitionType").add("PartitionColumnName")
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/TablesProcDir.java:52]: Put `PartitionType` at the end, so it is compatible with the previous version.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 162 / 178 (91.01%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/proc/PartitionsProcDir.java|29|34|85.29%|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|51|57|89.47%|
|:large_blue_circle:|com/starrocks/analysis/ListPartitionDesc.java|21|23|91.30%|
|:large_blue_circle:|com/starrocks/analysis/ShowPartitionsStmt.java|11|12|91.67%|
|:large_blue_circle:|com/starrocks/common/proc/TablesProcDir.java|50|52|96.15%|

File: fe/fe-core/src/main/java/com/starrocks/common/proc/TablesProcDir.java [hunk 10]
```diff
@@ -106,13 +110,13 @@ public ProcResult fetchResult() throws AnalysisException {
                 tableInfo.add(table.getId());
                 tableInfo.add(table.getName());
                 tableInfo.add(this.findIndexNum(tableType, table));
-                tableInfo.add(this.findPartitionType(tableType, table));
                 tableInfo.add(this.findPartitionKey(tableType, table));
                 tableInfo.add(this.findPartitionNum(tableType, table));
                 tableInfo.add(this.findState(tableType, table));
                 tableInfo.add(tableType);
                 tableInfo.add(TimeUtils.longToTimeString(table.getLastCheckTime()));
                 tableInfo.add(this.findRelicaCount(tableType, table));
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/common/proc/TablesProcDir.java:118]: findReplicaCount

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 162 / 178 (91.01%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/proc/PartitionsProcDir.java|29|34|85.29%|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|51|57|89.47%|
|:large_blue_circle:|com/starrocks/analysis/ListPartitionDesc.java|21|23|91.30%|
|:large_blue_circle:|com/starrocks/analysis/ShowPartitionsStmt.java|11|12|91.67%|
|:large_blue_circle:|com/starrocks/common/proc/TablesProcDir.java|50|52|96.15%|

**Review by wyb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wyb** [comment]
run starrocks_be_unittest

**Pull request merged**